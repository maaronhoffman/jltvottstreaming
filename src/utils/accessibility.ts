export const attachAccessibilityListener = () => {
  let isTabbing = false;

  // when the user presses the Tab key, we add a `is-tabbing` className to the body element, so we can enable outline
  // focus for keyboard users
  document.addEventListener('keydown', (event) => {
    if (event.key === 'Tab' && !isTabbing) {
      isTabbing = true;
      document.body.classList.add('is-tabbing');
    }
  });

  // remove the `is-tabbing` className when the user uses a mouse again to interact with elements
  document.addEventListener('mousedown', () => {
    if (isTabbing) {
      isTabbing = false;
      document.body.classList.remove('is-tabbing');
    }
  });
};
