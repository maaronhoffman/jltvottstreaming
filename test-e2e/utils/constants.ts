import { formatPrice } from './payments';

const baseUrl = 'http://localhost:8080/';

export const normalTimeout = 10;
export const longTimeout = 20;

export enum ShelfId {
  featured = 'featured',
  allFilms = 'all-films',
  allCourses = 'all-courses',
  continueWatching = 'continue_watching',
  favorites = 'favorites',
  liveChannels = 'live-channels',
}

export const makeShelfXpath = (shelf: ShelfId) => {
  return `//section[@data-testid="shelf-${shelf}"]`;
};

export default {
  username: 'ott-test-account@jwplayer.com',
  loginFormSelector: 'form[data-testid="login-form"]',
  registrationFormSelector: 'form[data-testid="registration-form"]',
  duplicateUserError: 'There is already a user with this email address',
  baseUrl: baseUrl,
  accountsUrl: `${baseUrl}u/my-account`,
  paymentsUrl: `${baseUrl}u/payments`,
  offersUrl: `${baseUrl}u/payments?u=choose-offer`,
  primitiveAnimalsId: 'oXGyKQ97',
  primitiveAnimalsTitle: 'Primitive Animals',
  primitiveAnimalsDescription: "If you're brand new to Blender or are getting stuck, check out the Blender 2.8 Fundamentals series.",
  agent327Title: 'Agent 327',
  agent327Description:
    'Hendrik IJzerbroot – Agent 327 – is a secret agent working for the Netherlands secret service agency. In the twenty comic books that were published since 1968, Martin Lodewijk created a rich universe with international conspiracies, hilarious characters and a healthy dose of Dutch humour.',
  bigBuckBunnyTitle: 'Big Buck Bunny',
  bigBuckBunnyDescription:
    "Big Buck Bunny (code-named Project Peach) is a 2008 short computer-animated comedy film featuring animals of the forest, made by the Blender Institute, part of the Blender Foundation. Like the foundation's previous film, Elephants Dream, the film was made using Blender, a free and open-source software application for 3D computer modeling and animation developed by the same foundation.",
  elephantsDreamTitle: 'Elephants Dream',
  elephantsDreamDescription:
    'Elephants Dream (code-named Project Orange during production and originally titled Machina) is a 2006 Dutch computer animated science fiction fantasy experimental short film produced by Blender Foundation using, almost exclusively, free and open-source software. The film is English-language and includes subtitles in over 30 languages.',
  fantasyVehicleTitle: 'Fantasy Vehicle Creation',
  fantasyVehicleId: 'lsFXY5xn',
  minecraftAnimationWorkshopTitle: 'Minecraft Animation Workshop',
  minecraftAnimationWorkshopDescription:
    'Ever wanted to learn how to make Minecraft Animation? Dillon walks you through the step by step process he used to create this Trident Animation. This includes everything from importing your Minecraft world to rendering and posting on YouTube. Enjoy!',
  startWatchingButton: 'Start watching',
  continueWatchingButton: 'Continue watching',
  signUpToWatch: 'Sign up to start watching!',
  continueWatchingShelfTitle: 'Continue Watching',
  paymentFields: {
    inplayer: {
      creditCardholder: 'Cardholder name',
      cardNumber: 'directPostCreditCardNumber',
      expiryDate: 'directPostExpiryDate',
      securityCode: 'directPostSecurityCode',
    },
    cleeng: {
      creditCardholder: 'Name on card',
      cardNumber: 'adyen-checkout__field--cardNumber',
      expiryDate: 'adyen-checkout__field--expiryDate',
      securityCode: 'adyen-checkout__field--securityCode',
    },
  },
  customRegFields: {
    topContainerSelector: 'div[data-testid="custom-reg-fields"]',
    termsAndConditionsField: 'input[type="checkbox"][name="terms"]',
    crfTextInput: '[data-testid="crf-input"]',
    crfCheckbox: '[data-testid="crf-checkbox"]',
    crfCountrySelect: '[data-testid="crf-country"]',
    crfUsStateSelect: '[data-testid="crf-us_state"]',
    crfDropdownSelect: '[data-testid="crf-select"]',
    crfRadioBox: '[data-testid="crf-radio"]',
    crfDateField: '[data-testid="crf-datepicker"]',
  },
  offers: {
    monthlyOffer: {
      cleeng: {
        label: `label[for="S970187168_NL"]`,
        price: formatPrice(6.99, 'EUR', 'NL'),
        paymentFee: formatPrice(0, 'EUR', 'NL'),
      },
      inplayer: {
        label: `label[for="S118699_38279"]`,
        price: formatPrice(6.99, 'EUR'),
        paymentFee: formatPrice(0, 'EUR'),
      },
    },
    yearlyOffer: {
      cleeng: {
        label: `label[for="S467691538_NL"]`,
        price: formatPrice(50, 'EUR', 'NL'),
        paymentFee: formatPrice(0, 'EUR', 'NL'),
      },
      inplayer: {
        label: `label[for="S118699_38280"]`,
        price: formatPrice(50, 'EUR'),
        paymentFee: formatPrice(0, 'EUR'),
      },
    },
  },
  creditCard: {
    inplayer: '4111111111111111',
    cleeng: '5555444433331111',
  },
};
