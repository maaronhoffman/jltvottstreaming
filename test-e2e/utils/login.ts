export function tryToSubmitForm(I: CodeceptJS.I) {
  I.submitForm(false);
  I.seeElementInDOM('div[class*=formFeedback]'); // This element can be visually hidden through CSS
  I.dontSee('Unknown error');
  I.dontSee('Incorrect email/password combination');
}

export function fillAndCheckField(I: CodeceptJS.I, field, value, error: string | boolean = false) {
  if (value === '') {
    // For some reason the Codecept/playwright clear and fillField with empty string do not fire the change events
    // so use key presses to clear the field to avoid test-induced bugs
    I.click(`input[name="${field}"]`);
    I.pressKey(['CommandOrControl', 'a']);
    I.pressKey('Backspace');
  } else {
    I.fillField(field, value);
  }

  checkField(I, field, error);
}

export async function checkField(I: CodeceptJS.I, field, error: string | boolean = false) {
  const hoverColor = 'rgba(255, 255, 255, 0.7)';
  const activeColor = error ? 'rgb(255, 53, 53)' : 'rgb(255, 255, 255)';
  const restingColor = error ? 'rgb(255, 53, 53)' : 'rgba(255, 255, 255, 0.34)';

  // If error === true, there's an error, but no associated message
  if (error && error !== true) {
    const helperId = await I.grabAttributeFrom(`input[name="${field}"]`, 'aria-describedby');
    I.see(error, `[data-testid=login-${field}-input]`);
    I.seeElement(`#${helperId}[class*=helperText]`);
    I.seeAttributesOnElements(`[name="${field}"]`, { 'aria-invalid': 'true' });
    I.seeCssPropertiesOnElements(`[data-testid="login-${field}-input"] [class*=helperText]`, { color: '#FF3535' });
  } else {
    I.dontSeeElement(`[class*=helperText] [data-testid="${field}-input"]`);
  }

  // There are 3 css states for the input fields, hover, active, and 'resting'. Check all 3.
  // This isn't so much for testing functionality, as it is to avoid test bugs caused by codecept placing the mouse
  // different places and accidentally triggering the wrong css color
  // Hover:
  I.click(`input[name="${field}"]`);
  I.seeCssPropertiesOnElements(`[data-testid="login-${field}-input"] [class*=container]`, { 'border-color': hoverColor });
  // Active (no hover):
  I.moveCursorTo('button[type=submit]');
  I.seeCssPropertiesOnElements(`[data-testid="login-${field}-input"] [class*=container]`, { 'border-color': activeColor });
  // Resting:
  I.click('div[class*=banner]');
  I.seeCssPropertiesOnElements(`[data-testid="login-${field}-input"] [class*=container]`, { 'border-color': restingColor });
}
