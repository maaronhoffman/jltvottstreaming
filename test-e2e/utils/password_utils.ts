import assert from 'assert';

export interface LoginContext {
  email: string;
  password: string;
}

async function testPasswordToggling(I: CodeceptJS.I, name = 'password') {
  await I.enableClipboard();
  await I.writeClipboard('');

  I.fillField({ name }, 'password123!');

  await checkPasswordType(I, name, 'password');

  I.click(`input[name="${name}"]+div button[aria-label="View password"]`);
  await checkPasswordType(I, name, 'text');

  await I.writeClipboard('dummy');

  I.click(`input[name="${name}"]+div button[aria-label="View password"]`);
  await checkPasswordType(I, name, 'password');
}

async function checkPasswordType(I: CodeceptJS.I, name, expectedType) {
  assert.strictEqual(await I.grabAttributeFrom(`input[name="${name}"]`, 'type'), expectedType);
}

export default {
  testPasswordToggling,
  createRandomEmail: function () {
    return `dummy-${Date.now()}-${Math.floor(Math.random() * 10 ** 6)}@jwplayer.com`;
  },
  createRandomPassword: function () {
    return `ABCDefgh${Math.floor(Math.random() * 10 ** 12)}!`;
  },
};
