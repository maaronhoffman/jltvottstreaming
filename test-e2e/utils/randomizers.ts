const random = (min = 0, max = 0) => min + Math.round(Math.random() * (max - min));

const randomDay = () => random(1, 28);

const randomMonth = () => random(1, 12);

const randomYear = () => random(2050, 2100);

export const randomDate = () => [randomDay(), randomMonth(), randomYear()];
