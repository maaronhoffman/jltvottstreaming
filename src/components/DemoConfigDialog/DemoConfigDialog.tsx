import React, { type ChangeEventHandler, type MouseEventHandler, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { type NavigateFunction, useNavigate } from 'react-router';
import { Helmet } from 'react-helmet';
import { createURL } from '@jwp/ott-common/src/utils/urlFormatting';
import { CONFIG_QUERY_KEY } from '@jwp/ott-common/src/constants';
import ErrorPage from '@jwp/ott-ui-react/src/components/ErrorPage/ErrorPage';
import TextField from '@jwp/ott-ui-react/src/components/form-fields/TextField/TextField';
import Button from '@jwp/ott-ui-react/src/components/Button/Button';
import ConfirmationDialog from '@jwp/ott-ui-react/src/components/ConfirmationDialog/ConfirmationDialog';
import LoadingOverlay from '@jwp/ott-ui-react/src/components/LoadingOverlay/LoadingOverlay';
import DevStackTrace from '@jwp/ott-ui-react/src/components/DevStackTrace/DevStackTrace';
import type { BootstrapData } from '@jwp/ott-hooks-react/src/useBootstrapApp';
import { AppError } from '@jwp/ott-common/src/utils/error';

import styles from './DemoConfigDialog.module.scss';

const regex = /^[a-z,\d]{0,8}$/g;
const DEMO_CONFIG = '225tvq1i';

interface State {
  configSource: string | undefined;
  error: string | Error | undefined;
  showDialog: boolean;
  loaded: boolean;
}

const initialState: State = {
  configSource: undefined,
  error: undefined,
  showDialog: false,
  loaded: false,
};

export function getConfigNavigateCallback(navigate: NavigateFunction) {
  return (configSource: string) => {
    navigate(
      {
        pathname: '/',
        search: new URLSearchParams([[CONFIG_QUERY_KEY, configSource]]).toString(),
      },
      { replace: true },
    );
  };
}

const DemoConfigDialog = ({ query }: { query: BootstrapData }) => {
  const { data, isLoading, error, refetch, isSuccess } = query;
  const { configSource: selectedConfigSource } = data || {};

  const { t } = useTranslation('demo');
  const navigate = useNavigate();
  const navigateCallback = getConfigNavigateCallback(navigate);

  const [state, setState] = useState<State>(initialState);

  const errorTitle = error && error instanceof AppError ? error.payload.title : '';
  const errorDescription = error && error instanceof AppError ? error.payload.description : '';

  const configNavigate = async (configSource: string | undefined) => {
    setState((s) => ({ ...s, configSource: configSource, error: undefined }));

    if (!configSource) {
      setState((s) => ({ ...s, error: t('enter_a_value') }));
    }
    // If trying to fetch the same config again, use refetch since a query param change won't work
    else if (configSource === selectedConfigSource) {
      await refetch();
    }
    // Get a new config by triggering a query param change
    else {
      // Force a page refresh so that application (including services and controllers) gets fully reinitialized
      window.location.href = createURL(window.location.href, { [CONFIG_QUERY_KEY]: configSource });
    }
  };

  useEffect(() => {
    // Don't grab values from props when config source is unset or still loading
    if (!selectedConfigSource || isLoading) {
      return;
    }

    // Initialize the config source if it's not yet set (this happens at first load)
    setState((s) => ({ ...s, configSource: s.configSource ?? selectedConfigSource }));

    // If there's an error after loading is done, grab it to display it to the user
    if (error) {
      setState((s) => ({ ...s, showDialog: false, error }));
    }
  }, [selectedConfigSource, error, isLoading]);

  const clearConfig = () => {
    setState(initialState);
    navigateCallback('');
  };

  const isValidSource = (configSource: string) => configSource.match(regex)?.some((m) => m === configSource);

  const onChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    setState((s) => ({
      ...s,
      configSource: event.target.value,
      error: isValidSource(event.target.value || '') ? undefined : t('invalid_config_id_format'),
    }));
  };

  const submitClick: MouseEventHandler<HTMLButtonElement> = async (event) => {
    event.preventDefault();

    let error: string = '';

    if (state.configSource?.length !== 8) {
      error = t('enter_8_digits');
    } else if (!isValidSource(state.configSource)) {
      error = t('invalid_config_id_format');
    }

    if (error) {
      setState((s) => ({ ...s, error: error }));
    } else {
      await configNavigate(state.configSource);
    }
  };

  const cancelConfigClick: MouseEventHandler<HTMLButtonElement> = (event) => {
    event.preventDefault();

    setState({ configSource: '', error: undefined, loaded: false, showDialog: true });
  };

  const confirmDemoClick = async () => {
    await configNavigate(DEMO_CONFIG);
  };

  const cancelDemoClick = () => setState((s) => ({ ...s, showDialog: false }));

  // If the config loads, reset the demo ui state
  useEffect(() => {
    if (isSuccess) {
      setState(initialState);
    }
  }, [setState, isSuccess]);

  // If someone links to the app with a config query,
  // we want to show the normal spinner instead of the dialog while trying to load the first time
  // Config is only undefined when it's the first load attempt.
  if (isLoading && state.configSource === undefined) {
    return <LoadingOverlay />;
  }

  return (
    <>
      {!isSuccess && (
        <Helmet>
          <meta name="description" content="Enter the App Config ID" />
        </Helmet>
      )}
      {isSuccess && (
        <div className={styles.note}>
          <div>{t('currently_previewing_config', { configSource: selectedConfigSource })}</div>
          <Button variant="text" label={t('click_to_unselect_config')} onClick={clearConfig} />
        </div>
      )}
      {!isSuccess && (
        <div className={styles.configModal}>
          <ErrorPage
            title={errorTitle || t('app_config_not_found')}
            message={errorDescription}
            learnMoreLabel={t('app_config_learn_more')}
            helpLink={'https://docs.jwplayer.com/platform/docs/ott-create-an-app-config'}
            error={typeof state.error === 'string' ? undefined : state.error}
          >
            <form method={'GET'} target={'/'}>
              <TextField
                required
                disabled={isLoading}
                className={styles.maxWidth}
                maxLength={8}
                value={state.configSource || ''}
                placeholder={t('please_enter_config_id')}
                name={'app-config'}
                autoCapitalize={'none'}
                error={!!state.error}
                helperText={
                  state.error && (
                    <span>
                      {typeof state.error === 'string' ? state.error : state.error?.message}
                      <br />
                      <br />
                      <DevStackTrace error={typeof state.error === 'string' ? undefined : state.error} />
                    </span>
                  )
                }
                onChange={onChange}
              />
              <div className={styles.controls}>
                <Button label={t('submit_config_id')} type={'submit'} onClick={submitClick} disabled={isLoading} busy={isLoading && !state.showDialog} />
                <Button label={t('cancel_config_id')} onClick={cancelConfigClick} disabled={isLoading} />
              </div>
            </form>
            <p>{t('use_the_jwp_dashboard')}</p>
            <p>{t('demo_note')}</p>
          </ErrorPage>
        </div>
      )}
      <ConfirmationDialog
        open={state.showDialog}
        title={t('view_generic_demo')}
        body={t('viewing_config_demo')}
        onConfirm={confirmDemoClick}
        onClose={cancelDemoClick}
        busy={isLoading}
      />
    </>
  );
};

export default DemoConfigDialog;
