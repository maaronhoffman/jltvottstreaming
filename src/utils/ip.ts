import { IS_DEVELOPMENT_BUILD, IS_PREVIEW_MODE, IS_TEST_MODE } from '@jwp/ott-common/src/utils/common';

import { overrideIPCookieKey } from '#src/constants';

export function getOverrideIP() {
  if (!IS_TEST_MODE && !IS_DEVELOPMENT_BUILD && !IS_PREVIEW_MODE) {
    return undefined;
  }

  return document.cookie
    .split(';')
    .find((s) => s.trim().startsWith(`${overrideIPCookieKey}=`))
    ?.split('=')[1]
    .trim();
}
