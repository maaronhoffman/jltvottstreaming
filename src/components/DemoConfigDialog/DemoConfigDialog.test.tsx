import React from 'react';
import { renderWithRouter } from '@jwp/ott-ui-react/test/utils';

import DemoConfigDialog from './DemoConfigDialog';

describe('<DemoConfigDialog>', () => {
  test('renders and matches snapshot', () => {
    const query = {
      isSuccess: true,
      error: null,
      isLoading: false,
      data: { configSource: 'abcdefgh', config: {}, settings: {} },
      refetch: () => Promise.resolve(null),
    };

    // @ts-expect-error
    const { container } = renderWithRouter(<DemoConfigDialog query={query} />);

    expect(container).toMatchSnapshot();
  });

  test('renders and matches snapshot error dialog', () => {
    const query = {
      isSuccess: false,
      error: new Error('smth'),
      isLoading: false,
      data: { configSource: 'aaaaaaaa', config: {}, settings: {} },
      refetch: () => Promise.resolve(null),
    };

    // @ts-expect-error
    const { container } = renderWithRouter(<DemoConfigDialog query={query} />);

    expect(container).toMatchSnapshot();
  });
});
