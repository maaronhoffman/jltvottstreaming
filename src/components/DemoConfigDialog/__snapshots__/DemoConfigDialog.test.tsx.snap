// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<DemoConfigDialog> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_note_561a54"
  >
    <div>
      currently_previewing_config
    </div>
    <button
      class="_button_e29948 _default_e29948 _text_e29948"
      type="button"
    >
      <span>
        click_to_unselect_config
      </span>
    </button>
  </div>
</div>
`;

exports[`<DemoConfigDialog> > renders and matches snapshot error dialog 1`] = `
<div>
  <div
    class="_configModal_561a54"
  >
    <div
      class="_errorPage_d73633"
    >
      <div
        class="_box_d73633"
      >
        <img
          alt=""
          class="_image_d73633"
          src="/images/logo.png"
        />
        <h1
          aria-live="polite"
          class="_title_d73633"
        >
          app_config_not_found
        </h1>
        <div
          class="_main_d73633"
        >
          <p
            class="_message_d73633"
          >
            generic_error_description
          </p>
          <form
            method="GET"
            target="/"
          >
            <div
              class="_formField_a245e1 _textField_0fa545 _maxWidth_561a54"
            >
              <div
                class="_container_ac31ed _error_ac31ed"
              >
                <input
                  aria-describedby="helper_text_1235_app-config"
                  aria-invalid="true"
                  aria-required="true"
                  autocapitalize="none"
                  class="_input_ac31ed"
                  id="text-field_1235_app-config"
                  maxlength="8"
                  name="app-config"
                  placeholder="please_enter_config_id"
                  required=""
                  type="text"
                  value="aaaaaaaa"
                />
              </div>
              <div
                class="_helperText_edc1c5 _error_edc1c5"
                id="helper_text_1235_app-config"
              >
                <span>
                  smth
                  <br />
                  <br />
                  <a>
                    Show Stack Trace
                  </a>
                </span>
              </div>
            </div>
            <div
              class="_controls_561a54"
            >
              <button
                aria-disabled="false"
                class="_button_e29948 _default_e29948 _outlined_e29948"
                type="submit"
              >
                <span>
                  submit_config_id
                </span>
              </button>
              <button
                aria-disabled="false"
                class="_button_e29948 _default_e29948 _outlined_e29948"
                type="button"
              >
                <span>
                  cancel_config_id
                </span>
              </button>
            </div>
          </form>
          <p>
            use_the_jwp_dashboard
          </p>
          <p>
            demo_note
          </p>
          <div
            class="_links_d73633"
          >
            <a
              href="https://docs.jwplayer.com/platform/docs/ott-create-an-app-config"
              rel="noreferrer"
              target="_blank"
            >
              app_config_learn_more
            </a>
            <span
              class="_stack_d73633"
            >
              <a>
                Show Stack Trace
              </a>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
